# Working with APIs

## Objective

Working with APIs by finding the position of the International Space Station.

## Retrieve Data with Requests

In [1]:
import requests

# Make a get request to get the latest position of the international space station from the opennotify api.
response = requests.get("http://api.open-notify.org/iss-now.json")

# Print the status code of the response.
print(response.status_code)

200


    •	200 - Everything went okay, and the server returned a result (if any).

In [2]:
response = requests.get("http://api.open-notify.org/iss-pass")
print(response.status_code)

404


    •	404 - The server didn't find the resource you tried to access.

iss-pass wasn’t a valid endpoint; add .json at the end, as the [API documentation](http://open-notify.org/Open-Notify-API/) states.

In [3]:
response = requests.get("http://api.open-notify.org/iss-pass.json")
status_code = response.status_code
status_code

400

    •	400 - The server thinks you made a bad request. This can happen when you don't send the information the API requires to process your request, among other things.

Look at the documentation for the OpenNotify API, see that the ISS Pass endpoint requires two parameters.

In [6]:
# Set up the parameters we want to pass to the API.
# This is the latitude and longitude of New York City.
parameters = {"lat": 40.71, "lon": -74}

# Make a get request with the parameters.
response = requests.get("http://api.open-notify.org/iss-pass.json", params=parameters)

# Print the content of the response (the data the server returned)
print(response.content)

# This gets the same data as the command above
response = requests.get("http://api.open-notify.org/iss-pass.json?lat=40.71&lon=-74")
print(response.content)
print(type(response.content))

b'{\n  "message": "success", \n  "request": {\n    "altitude": 100, \n    "datetime": 1503629872, \n    "latitude": 40.71, \n    "longitude": -74.0, \n    "passes": 5\n  }, \n  "response": [\n    {\n      "duration": 558, \n      "risetime": 1503661140\n    }, \n    {\n      "duration": 641, \n      "risetime": 1503666867\n    }, \n    {\n      "duration": 576, \n      "risetime": 1503672720\n    }, \n    {\n      "duration": 550, \n      "risetime": 1503678583\n    }, \n    {\n      "duration": 616, \n      "risetime": 1503684388\n    }\n  ]\n}\n'
b'{\n  "message": "success", \n  "request": {\n    "altitude": 100, \n    "datetime": 1503629872, \n    "latitude": 40.71, \n    "longitude": -74.0, \n    "passes": 5\n  }, \n  "response": [\n    {\n      "duration": 558, \n      "risetime": 1503661140\n    }, \n    {\n      "duration": 641, \n      "risetime": 1503666867\n    }, \n    {\n      "duration": 576, \n      "risetime": 1503672720\n    }, \n    {\n      "duration": 550, \n      "r

## Working with JSON data

Noticed that the content of the response earlier was a string (although it was shown as a bytes object); convert the content to a string using response.content.decode("utf-8").

JSON is a way to encode data structures like lists and dictionaries to strings that ensures that they are easily readable by machines. JSON is the primary format in which data is passed back and forth to APIs, and most API servers will send their responses in JSON format.

### Getting JSON from an API request

In [5]:
import json

# Make the same request with the coordinates of San Francisco instead.
parameters = {"lat": 37.78, "lon": -122.41}
response = requests.get("http://api.open-notify.org/iss-pass.json", params=parameters)

# Get the response data as a python object. Verify that it's a dictionary.
data = response.json()
print(type(data))
print(data)

<class 'dict'>
{'message': 'success', 'request': {'altitude': 100, 'datetime': 1503630183, 'latitude': 37.78, 'longitude': -122.41, 'passes': 5}, 'response': [{'duration': 577, 'risetime': 1503672195}, {'duration': 633, 'risetime': 1503677942}, {'duration': 526, 'risetime': 1503683826}, {'duration': 475, 'risetime': 1503689718}, {'duration': 578, 'risetime': 1503695523}]}


## Content type

Show metadata containing information on how the data was generated and how to decode it.

In [7]:
# Headers is a dictionary
print(response.headers)

# Get the content-type from the dictionary.
print(response.headers["content-type"])

{'Server': 'nginx/1.10.3', 'Date': 'Fri, 25 Aug 2017 03:03:21 GMT', 'Content-Type': 'application/json', 'Content-Length': '519', 'Connection': 'keep-alive', 'Via': '1.1 vegur'}
application/json


## Finding the number of people in space

OpenNotify has one more API endpoint, astros.json. It tells how many people are currently in space.

In [9]:
# Get the response from the API endpoint.
response = requests.get("http://api.open-notify.org/astros.json")
data = response.json()

# 9 people are currently in space.
print(data["number"])
print(data)

6
{'message': 'success', 'people': [{'name': 'Peggy Whitson', 'craft': 'ISS'}, {'name': 'Fyodor Yurchikhin', 'craft': 'ISS'}, {'name': 'Jack Fischer', 'craft': 'ISS'}, {'name': 'Sergey Ryazanskiy', 'craft': 'ISS'}, {'name': 'Randy Bresnik', 'craft': 'ISS'}, {'name': 'Paolo Nespoli', 'craft': 'ISS'}], 'number': 6}
